examples/type_conversion.rs

```diff
@@ -0,0 +1,97 @@
+use approx::assert_abs_diff_eq;
+use ndarray::prelude::*;
+use std::convert::TryFrom;
+
+#[cfg(feature = "approx")]
+fn main() {
+    // Converting an array from one datatype to another is implemented with the
+    // `ArrayBase::mapv()` function. We pass a closure that is applied to each
+    // element independently. This allows for more control and flexiblity in
+    // converting types.
+    //
+    // Below, we illustrate three different approaches for the actual conversion
+    // in the closure.
+    // - `From` ensures perfect conversions known at compile time and is the
+    //   best default choice.
+    // - `TryFrom` either converts data perfectly or panics, ensuring that the
+    //   rest of the program does not continue with unexpected data.
+    // - `as` never panics and silently performs a saturating cast. It should
+    //   only be used for simple conversions.
+    //
+    // For a brush-up on casting between numeric types in Rust, refer to:
+    // https://doc.rust-lang.org/rust-by-example/types/cast.html
+
+    // Infallible, safe conversion with `From`
+    // The trait `std::convert::From` is only implemented for conversions that
+    // can be guaranteed to be safe at compile time. This is the safest
+    // approach.
+    let a_u8: Array<u8, _> = array![[1, 2, 3], [4, 5, 6]];
+    let a_f32 = a_u8.mapv(|element| f32::from(element));
+    assert_abs_diff_eq!(a_f32, array![[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]);
+
+    // Fallible, safe conversion with `TryFrom`
+    // `i8` numbers can be negative, in such a case, there is no perfect
+    // conversion to `u8` defined. In this example, all numbers are positive and
+    // in bounds and can be converted at runtime. But for unknown runtime input,
+    // this would panic with the message provided in `.expect()`. Note that you
+    // can also use `.unwrap()` to be more concise.
+    let a_i8: Array<i8, _> = array![120, 8, 0];
+    let a_u8 = a_i8.mapv(|element| u8::try_from(element).expect("Could not convert i8 to u8"));
+    assert_eq!(a_u8, array![120u8, 8u8, 0u8]);
+
+    // Simple upcasting with `as`
+    // Every `u8` fits perfectly into a `u32`, therefore we do not need to take
```

Comment:
In realistic programs, one often starts with this reasoning but especially with type interference the involved types can change and `as` becomes lossy even if it started out lossless. Hence, I think `From` should be recommended whenever possible. If the types changes and cannot be converted losslessly any more, compilation will fail and one can still decided to "downgrade" to a cast.

Reply:
I think it's good to have examples with as, so we should have some

Reply:
I do agree that there should be examples, but I would recommend against choosing one where `From` could be used instead. The `as` example should rather do a potentially lossy conversion that cannot be done using `From` probably with a comment to use this if `TryFrom` is a performance issue or does not apply.

Reply:
Hm the prime example for me would be `f32 as i32` - but already have an example from floats to integers to show the saturating cast. With the updated comments/documentation, do you think it is fine or which example would you like to see in particular @adamreichold ?

Reply:
> do you think it is fine or which example would you like to see in particular @adamreichold ?

I think it is fine. The other one rather common example (for example when working with integer coordinates on a grid) is between signed and unsigned types, e.g. `usize as isize` (which can fail as `usize::MAX > isize::MAX as usize`).

Reply:
Ah thank you for the example, I included it in `type_conversion.rs` :slightly_smiling_face: